opm/simulators/flow/TracerModel.hpp

```diff
@@ -348,31 +464,112 @@ class TracerModel : public GenericTracerModel<GetPropType<TypeTag, Properties::G
         const auto& eclWell = well.wellEcl();
         for (int tIdx = 0; tIdx < tr.numTracer(); ++tIdx) {
             this->wellTracerRate_[std::make_pair(eclWell.name(), this->name(tr.idx_[tIdx]))] = 0.0;
+            this->wellFreeTracerRate_[std::make_pair(eclWell.name(), this->wellfname(tr.idx_[tIdx]))] = 0.0;
+            this->wellSolTracerRate_[std::make_pair(eclWell.name(), this->wellsname(tr.idx_[tIdx]))] = 0.0;
         }
 
-        std::vector<double> wtracer(tr.numTracer());
+        std::vector<Scalar> wtracer(tr.numTracer());
         for (int tIdx = 0; tIdx < tr.numTracer(); ++tIdx) {
             wtracer[tIdx] = this->currentConcentration_(eclWell, this->name(tr.idx_[tIdx]));
         }
 
-        for (auto& perfData : well.perforationData()) {
-            const auto I = perfData.cell_index;
+        Scalar dt = simulator_.timeStepSize();
+        std::size_t well_index = simulator_.problem().wellModel().wellState().index(well.name()).value();
+        const auto& ws = simulator_.problem().wellModel().wellState().well(well_index);
+        for (std::size_t i = 0; i < ws.perf_data.size(); ++i) {
+            const auto I = ws.perf_data.cell_index[i];
             Scalar rate = well.volumetricSurfaceRateForConnection(I, tr.phaseIdx_);
-            if (rate > 0) {
+            Scalar rate_s;
+            if (tr.phaseIdx_ == FluidSystem::oilPhaseIdx && FluidSystem::enableVaporizedOil()) {
+                rate_s = ws.perf_data.phase_mixing_rates[i][ws.vaporized_oil];
+            }
+            else if (tr.phaseIdx_ == FluidSystem::gasPhaseIdx && FluidSystem::enableDissolvedGas()) {
+                rate_s = ws.perf_data.phase_mixing_rates[i][ws.dissolved_gas];
+            }
+            else {
+                rate_s = 0.0;
+            }
+
+            Scalar rate_f = rate - rate_s;
+            if (rate_f > 0) {
```

Comment:
Can we get the free rate from ws.perf_data directly?

Reply:
I looked but couldn't find any at least.